### PR TITLE
Disable EDNS DO

### PR DESCRIPTION
**Pull request by wcawijngaards**
Disable the EDNS DO flag in upstream requests. This can be helpful for devices that cannot handle DNSSEC information. But it should not be enabled otherwise, because that would stop DNSSEC validation. The DNSSEC validation would not work for Unbound itself, and also not for downstream users. Default is no.

The option is `disable-edns-do: no`.

### CODE REVIEW

File: validator/validator.c [hunk 1]
```diff
@@ -200,6 +200,17 @@ val_init(struct module_env* env, int id)
 		log_err("validator: could not apply configuration settings.");
 		return 0;
 	}
+	if(env->cfg->disable_edns_do) {
+		struct trust_anchor* anchor = anchors_find_any_noninsecure(
+			env->anchors);
+		if(anchor) {
+			char b[257];
```
Comment (gthess) [inline validator/validator.c]: I would prefer the use of a defined valued instead
```suggestion
			char b[LDNS_MAX_DOMAINLEN+2];
```

File: doc/unbound.conf.5.in [hunk 2]
```diff
@@ -1302,6 +1302,15 @@ servers that set the CD flag but cannot validate DNSSEC themselves are
...
 The default value is "no".
 .TP
+.B disable\-edns\-do: \fI<yes or no>
+Disable the EDNS DO flag in upstream requests. This can be helpful for
+devices that cannot handle DNSSEC information. But it should not be enabled
+otherwise, because that would stop DNSSEC validation. The DNSSEC validation
+would not work for Unbound itself, and also not for downstream users.
+When the option is enabled, queriers that set the DO flag receive no EDNS
+record in the response to indicate the lack of support to them.
+Default is no.
```
Comment (gthess) [inline doc/unbound.conf.5.in]: With the newer commits I find the following text more clear.
```suggestion
Disable the EDNS DO flag in upstream requests.
It breaks DNSSEC validation for Unbound's clients.
This results in the upstream name servers to not include DNSSEC records in
their replies and could be helpful for devices that cannot handle DNSSEC
information.
When the option is enabled, clients that set the DO flag receive no EDNS
record in the response to indicate the lack of support to them.
If this option is enabled but Unbound is already configured for DNSSEC
validation (i.e., the validator module is enabled; default) this option is
implicitly turned off with a warning as to not break DNSSEC validation in
Unbound.
Default is no.
```

**Review by gthess — APPROVED** [review]
The changes look good to me.
I am not certain about the name of the option since it is very technical, but then again this option should only be considered if you actually know what you are doing and the consequences to the resolver's clients. So maybe it is fine this way; also I can't come up with a better name.

File: validator/validator.c (hunk 1 above)
Comment (wcawijngaards) [reply validator/validator.c]: Thanks, committed.

File: doc/unbound.conf.5.in (hunk 2 above)
Comment (wcawijngaards) [reply doc/unbound.conf.5.in]: Okay, sure, committed the new text.

**Review by Philip-NLnetLabs — APPROVED** [review]

**Pull request merged**